src/main/scala/de/m7w3/signal/DeviceRegistration.scala

```diff
@@ -2,72 +2,125 @@ package de.m7w3.signal
+
+      def oneFieldEmpty(): Boolean = deviceName.text.value.trim.isEmpty || userId.text.value.trim.isEmpty || dbPassword.text.value.trim.isEmpty
+
       ok.onAction = (a: ActionEvent) => {
-        Step2.setVisible(false)
+        this.setVisible(false)
         Step3.setVisible(true)
         // TODO: do this in a better way
         Main.account = AccountHelper(userId.getText())
+        //TODO: show a progress bar while the future is not complete
         Platform.runLater{
-          val outputstream = QRCodeGenerator.generate(() => Main.account.getNewDeviceURL)
-          val image = new Image(new ByteArrayInputStream(outputstream.toByteArray), 600D, 600D, true, true)
-          val backgroundImage = new BackgroundImage(image, BackgroundRepeat.NoRepeat, BackgroundRepeat.NoRepeat, BackgroundPosition.Center, BackgroundSize.Default)
-          Step3.qrCode.setBackground(new Background(Array(backgroundImage)))
+          println("pressed")
+          Future(QRCodeGenerator.generate(() => Main.account.getNewDeviceURL)).map(outputstream => {
+            val image = new Image(new ByteArrayInputStream(outputstream.toByteArray), 300D, 300D, true, true)
+            // Step3.add(new ImageView(image), 0, 1)
+            val backgroundImage = new BackgroundImage(image, BackgroundRepeat.NoRepeat, BackgroundRepeat.NoRepeat, BackgroundPosition.Center, BackgroundSize.Default)
+            Step3.qrCode.setBackground(new Background(Array(backgroundImage)))
+            Step3.finish.disable = false
+          })
         }
       }
-      this.children = List(deviceLable, deviceName, userIdLable, userId, ok)
+      ok.defaultButtonProperty().bind(dbPassword.focusedProperty())
+
+      alignment = Pos.Center
+      visible = false
+      hgap = 10
+      vgap = 20
+      padding = Insets(20, 150, 10, 10)
+      add(new Label("Device name:"), 0, 0)
+      add(deviceName, 1, 0)
+      add(new Label("Phone number: "), 0, 1)
+      add(userId, 1, 1)
+      add(new Label("Password:"), 0, 2)
+      add(dbPassword, 1, 2)
+      add(ok, 0, 3)
     }
 
-    object Step3 extends VBox{
-      Step3.setVisible(false)
-      this.alignment = Pos.Center
-      private val label = new Label("Open Signal on your phone and go to Settings > Devices. Click on the plus icon and scan the above QR code.")
+    object Step3 extends GridPane{
       val qrCode = new StackPane
-      qrCode.prefWidth = 600D
-      qrCode.prefHeight = 600D
+      qrCode.prefWidth = 300D
+      qrCode.prefHeight = 300D
       val finish = new Button("Finish")
-      finish.onAction = (a: ActionEvent) => {
-        Platform.runLater{
-          Main.store = context.createNewProtocolStore(Step2.dbPassword.getText())
-          Main.account.finishDeviceLink(Step2.deviceName.getText(), Main.store)
+      finish.disable = true
+      finish.onAction = (a: ActionEvent) => Platform.runLater{
+        // TODO show a progress bar
+        Main.store = context.createNewProtocolStore(Step2.dbPassword.getText())
```

Comment:
store should be a property of ApplicationContext or something along those lines, same for account